### PR TITLE
gcp authentication integration

### LINKED ISSUE #2213: Support using GCP account to access Chaos Dashboard.

### PR DESCRIPTION
**Pull request by YangKeao**
### What problem does this PR solve?

Issue Number: close #2213

Problem Summary:

### What is changed and how it works?

What's Changed:

It provides three endpoints and adds a middleware:

```
GET /api/auth/gcp/redirect
```

It will redirect to the google login.

```
GET /api/auth/gcp/callback
```

It will exchange the oauth `access_token` and `refresh_token`. These token and the expiry will be included in the cookie. **notice:** the expiry will be URLEncoded (as in the cookie).

A middleware is added to the `gin` server, so that when the `x-authentication-method` is set to `gcp`, it will read token from the header and set the automatically refreshed token to the `Authentication` header.

It's waiting for the frontend change, or should we merge the PR first?

### CODE REVIEW

**Comment by STRRL** [comment]
please update the `install.sh`

**Review by cwen0 — COMMENTED** [review]
Rest LGTM

File: pkg/dashboard/apiserver/auth/gcp/service.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+	c.Redirect(http.StatusFound, uri)
+}
+
+func (s *Service) authCallback(c *gin.Context) {
+	ctx := c.Request.Context()
+
+	oauth := s.getOauthConfig(c)
+	oauth2Token, err := oauth.Exchange(ctx, c.Request.URL.Query().Get("code"), oauth2.AccessTypeOffline, oauth2.ApprovalForce)
+	log.Info("get oauth2Token", "oauth2Token", oauth2Token)
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply pkg/dashboard/apiserver/auth/gcp/service.go:94]: Nice catch!

[the remark this answers is not in the document]
Comment (YangKeao) [reply pkg/dashboard/apiserver/auth/gcp/service.go]: Fixed.

**Review by cwen0 — APPROVED** [review]
LGTM

**Review by g1eny0ung — CHANGES_REQUESTED** [review]

File: ui/package.json [hunk 2]
```diff
@@ -24,6 +24,7 @@
     "flat": "^5.0.2",
     "formik": "^2.2.6",
     "idb": "^6.1.2",
+    "js-cookie": "^3.0.0",
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply ui/package.json]: `/ui` folder has been reset to `upstream/master`.

File: pkg/dashboard/apiserver/auth/gcp/middleware.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	log.Info("handling gcp middleware")
+	if c.Request.Header.Get("X-Authorization-Method") != "gcp" {
+		c.Next()
+		return
+	}
+
+	expiry, err := time.Parse(time.RFC3339, c.Request.Header.Get("X-Authorization-Expiry"))
+	if err != nil {
+		c.Status(http.StatusInternalServerError)
+		_ = c.Error(utils.ErrInternalServer.WrapWithNoMessage(err))
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply pkg/dashboard/apiserver/auth/gcp/middleware.go]: Fixed.

**Comment by YangKeao** [comment]
@g1eny0ung PTAL

**Review by g1eny0ung — APPROVED** [review]
LGTM

**Review by cwen0 — APPROVED** [review]

**Pull request merged**